### ISSUE TITLE
add dynamic support for the container constraints for the SwarmSpawner

### ISSUE THREAD

**Issue by drzhouq** [jupyterhub/dockerspawner#440]
### Proposed change

One can specify "static" constraint for the container (for example, c.SwarmSpawner.extra_placement_spec = { 'constraints' : ['node.role==worker'] }) to be spawn by the jupyterhub server. Is it possible to dynamic change the constraint based on image chosen by the user or even the user name?

### Who would use this feature?

Some images need to use GPU, and not every computer in a swarm has GPU. Enable dynamic constraints for the container will help to ensure the container spawn can have access to the GPU.

### (Optional): Suggest a solution

It seems that c.Spawner.pre_spawn_hook is currently too late to apply the constraint to container spawn, but it should be the logic place to implement this "dynamic" container constraint.

**Comment by minrk** [comment]
Indeed, most of these `extra_` fields should be allowed to be callables. Starting on this in #500, swarm should come next
